kafka/main.go

```diff
@@ -67,82 +82,165 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
+							msg,
+							topicMapping.ResultProducer,
+							providedServer,
+							logger,
+						)
+						if err != nil {
+							logger.Error().Err(err).Msg("Processing failed.")
+							errorResults <- err
 						}
+
+					}(msg, topicMapping, providedServer, logger)
+				}
+			}
+			// If the topic in the message doesn't match andy of the topicMappings
+			// then the goroutine will not be spawned and wg.Done() won't be
+			// called. If this happens, be sure to call it.
+			if !wgDoneDeferred {
+				wg.Done()
+			}
+		}
+		close(errorResults)
+		// Iterate over any failures and create an error slice
+		var temporaryErrors []*utils.ProcessingError
+		for processingError := range errorResults {
+			if processingError.Temporary {
+				temporaryErrors = append(temporaryErrors, processingError)
+			} else {
+				continue
+			}
+		}
+
+		// If there are temporary errors, sort them so that the first item in the
+		// list has the lowest offset. Only run sort if there is more than one
+		// temporary error.
+		if len(temporaryErrors) > 0 {
+			logger.Error().Msgf("temporary errors: %#v", temporaryErrors)
+			if len(temporaryErrors) > 1 {
+				sort.Slice(temporaryErrors, func(i, j int) bool {
+					return temporaryErrors[i].KafkaMessage.Offset < temporaryErrors[j].KafkaMessage.Offset
+				})
+			}
+			// Iterate over the batch to find the message that came before the
+			// first temporary failure and commit it. This will ensure that
+			// the temporary failure is picked up as the first item in the next
+			// batch.
+			for _, message := range batch {
+				if message.Offset == temporaryErrors[0].KafkaMessage.Offset-1 {
+					if err := reader.CommitMessages(ctx, message); err != nil {
+						logger.Error().Msgf("failed to commit: %s", err)
 					}
+					// Before retrying the temporary failure, wait the
+					// prescribed time.
+					time.Sleep(temporaryErrors[0].Backoff)
 				}
 			}
-		}(topicMappings)
+			// If there are no temporary errors sort the batch in descending order by
+			// offset and then commit the offset of the first item in the list.
+		} else if len(batch) > 0 {
+			sort.Slice(batch, func(i, j int) bool {
+				return batch[i].Offset < batch[j].Offset
+			})
+			logger.Info().Msgf("Committing offset", batch[0].Offset)
+			if err := reader.CommitMessages(ctx, batch[0]); err != nil {
+				logger.Error().Err(err).Msg("failed to commit")
+			}
+		}
 	}
+}
 
-	return nil
+// Pull messages out of the Reader's underlying batch so that they can be processed in parallel
+// There is an ongoing discussion of batch message processing implementations with this
+// library here: https://github.com/segmentio/kafka-go/issues/123
+func batchFromReader(ctx context.Context, reader *kafka.Reader, count int, logger *zerolog.Logger) ([]kafka.Message, error) {
+	var (
+		messages []kafka.Message
+		err      error
+	)
+	for i := 0; i < count; i++ {
+		innerctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+		defer cancel()
```

Comment:
I was thinking that this would allow us to time out individual processing attempts, but that isn't currently implemented. I'm happy to put it outside the loop if you think that would be better. I've done that and you can share further thoughts if you think they're needed.